### ISSUE TITLE
Support for Confidential Clients with the Authorization Code Flow

### ISSUE THREAD

**Issue by DJDANNY123** [openauthjs/openauth#96]
[The Authorization Code Flow in OAuth is optimised for Confidential OAuth clients](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1).

From my understanding, the main benefit of using a confidential client for this, is so that an authorization code cannot be hijacked and used by another application/actor which knows/shares the `clientId` of a public client.

[Auth0 appears to go as far as to only allow this flow with confidential clients due to this security implication.](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow)

Seems somewhat related to [this issue relating to Resource Owner Password flow](https://github.com/openauthjs/openauth/issues/18) as they both require support for confidential OAuth clients, but I think this flow is preferred or seen as better practice.

**Comment by pverheggen** [comment]
I believe this is already supported, the `"code"` argument tells it to use Authorization Code Flow and the `pkce` option determines whether to add PKCE or not.

```js
client.authorize(redirectUrl, "code", { pcke: false })
```

If you look in the examples folder, `react` is an example of a public client that uses PKCE, and `cloudflare-api`/ `lambda-api` are confidential clients that do not.

**Comment by bchilcott** [comment]
@pverheggen I think the question is about whether [confidential clients](https://datatracker.ietf.org/doc/html/rfc6749#section-2.1), as in clients with both an ID and secret, are likely to be supported. I haven't been able to find any concrete explanation of whether using pkce is sufficient to replace the need for confidential clients.

**Comment by pverheggen** [comment]
If I understand correctly - those examples are still considered "confidential clients" since they run on the server, but the problem is that they're supposed to be sending client ID/secret when doing the code exchange, and currently not doing that.  

Do I have that right?

**Comment by bchilcott** [comment]
The examples you linked are using public clients, since they are not sending a secret to the server along with the client ID. 

From my understanding of the oauth spec, the auth server should hold a record of clients and their secrets, and validate that each request from the client includes its respective secret.

**Comment by pverheggen** [comment]
Gotcha, thanks for clarifying.  I was thrown off by the following definition:

>    confidential
>       Clients capable of maintaining the confidentiality of their
>       credentials (e.g., client implemented on a secure server with
>       restricted access to the client credentials), or capable of secure
>       client authentication using other means.

They are certainly _capable_ of maintaining confidentiality, it just isn't implemented.  

Regarding PKCE, my understanding is that it's the next best thing when client confidentiality is not possible, so I agree it doesn't replace the need for this.

**Comment by bchilcott** [comment]
That threw me off too, there's some better clarification on confidential clients [here](https://datatracker.ietf.org/doc/html/rfc6749#section-2.3).

**Comment by thdxr** [comment]
yeah so i omitted requiring client_secret on code request initially because it's not strictly needed for the initial flows we supported.

that said i've been thinking about reusing the `allow` callback to let you also verify that an exchange is allowed to happen - you can verify the client_id with a client_secret they passed in

at some point maybe the idea of client will be a native thing and stored in storage - just unclear right now if that's a good move

**Comment by bchilcott** [comment]
I wouldn't be opposed to managing the storage of clients outside OpenAuth and just verifying them in the allow callback, at least at first. As long as `client` can send the secret to be used in the `allow` callback.
